recipe-client-addon/lib/NormandyDriver.jsm

```diff
@@ -156,6 +156,68 @@ this.NormandyDriver = function(sandboxManager) {
       sandboxManager.removeHold(`setTimeout-${token}`);
     },
 
+    /*
+     * Return a promise that resolves to an Addon ID if installation is successful.
+     */
+    installAddon(url) {
+        /* Install an addon.  Note that the AddonManager can only handle
+         * file:// and downloadable URLs.
+         */
+
+      var installObjectCallback = (installObject, resolve, reject) => {
+        installObject.install();
+        installObject.addListener({
+          onInstallEnded: (addonInstall, addon) => {
+            let addonId = addon.id;
+            resolve(addonId);
+          },
+          onInstallFailed: (addonInstall) => {
+            reject(`AddonInstall error code: [${addonInstall.error}]`);
+          },
+        });
+      };
+
+      function installObjectPromise(installObject) {
+        return new Promise(function(resolve, reject) {
+          installObjectCallback(installObject, resolve, reject);
+        });
+      }
+
+      function getInstallForURLPromise(url) {
+        return new Promise(function(resolve, reject) {
+          AddonManager.getInstallForURL(url, resolve, 'application/x-xpinstall');
+        });
+      };
+
+      // Return a promise
+      return getInstallForURLPromise(url).then(installObjectPromise);
+    },
+
+    /*
+     * Return a promise that resolves to a success messsage if
+     * addon uninstall is successful.
+     */
+    uninstallAddon(addonId) {
+      function uninstallPromise(id) {
+        return new Promise(function(resolve, reject) {
+          AddonManager.getAddonByID(id, addon => {
+            if (addon != null) {
+              try {
+                addon.uninstall();
+              } catch (e) {
+                reject("Addon uninstall triggered an error");
+                return;
+              }
+            } else {
+              reject("Addon is null - can't uninstall it");
+            }
+            resolve();
+          });
+        });
+      }
+      return uninstallPromise(addonId);
```

Comment:
I changed this to be an immediately invoked function as I need to capture the addonId within the Promise using a closure.